workspaces/cli-server/src/spectacle/index.ts

```diff
@@ -109,6 +145,38 @@ export class LocalCliSpecRepository implements IOpticSpecReadWriteRepository {
       return events;
     }
   }
+
+  async resetToCommit(batchCommitId: string): Promise<void> {
+    // TODO handle partial error + cleanup
+    const specJsFile = path.join(
+      this.dependencies.specDirPath,
+      'specification.json'
+    );
+    const tempOutputFile = path.join(
+      this.dependencies.specDirPath,
+      `temp-spec-file${uuidv4()}.json`
+    );
+    const tempFileDeletion = path.join(
+      this.dependencies.specDirPath,
+      `spec-to-delete${uuidv4()}.json`
+    );
```

Comment:
I'm not sure if this is necessary? But if node crashes while the stream is in flight, this could (maybe) corrupt the specification.json file?

Is there a better way to handle this?

Reply:
Ensuring some safety of the data is a good thing and the process going away (killed, crashing, etc) an hopefully infrequent but plausible situation we must consider, nice that you've thought about this. 

We're in luck, though: in this case, `specification.json` that has any historical value is tracked in their Git repo, so it's hard to really destroy something. Git is likely to hold at least _some_ version for us to mitigate the damage caused by crashes. I'd say that's probably enough of a safety net for now, given the appetite for this bet and it's fairly experimental nature.

What if we _do_ want to increase safety some more?

Renaming files seems like a reasonable way to do so: the chance of data getting lost is definitely minimised this way, as data isn't destroyed until the previous operation has written. However, it does introduce a whole bunch more failure cases as well that could lead to interesting situations. What happens, for instance, when the process crashes _after_ writing the tmp output file, but _before_ it can rename it? Or before the temp file can be deleted? How should the process recover, or are we stuck with old, potentially partly written files?

Generally, the way to deal with these kinds of problems is to implement a crash-recovery mechanism, or a processing mechanism that can continue where it left off. Doing that efficiently gets you into the world or Write-Ahead-Logs or similar concepts: basically we're implementing a database at that point.

There are some other routes to look into, like using `fs.truncate`, but to make that work you have to be pretty precise as well, so there's no guarantees that actually makes Optic more instead of less fault-tolerant.

For now I'd say: lets rely on the fact that these specs are tracked files in Git repositories as our safety net: the rest of the app definitely does so. If we _do_ want to start improving fault tolerance beyond that, I'd say we should look at implementing that sort of stuff on the Rust side.

Reply:
Ah yeah, it completely slipped my mind that anything of value would most likely be backed by version control, so it's good enough to rely on that rather than build our own crash recovery mechanisms